beacon-chain/core/altair/block.go

```diff
@@ -0,0 +1,137 @@
+//    total_base_rewards = Gwei(get_base_reward_per_increment(state) * total_active_increments)
+//    max_participant_rewards = Gwei(total_base_rewards * SYNC_REWARD_WEIGHT // WEIGHT_DENOMINATOR // SLOTS_PER_EPOCH)
+//    participant_reward = Gwei(max_participant_rewards // SYNC_COMMITTEE_SIZE)
+//    proposer_reward = Gwei(participant_reward * PROPOSER_WEIGHT // (WEIGHT_DENOMINATOR - PROPOSER_WEIGHT))
+//
+//    # Apply participant and proposer rewards
+//    all_pubkeys = [v.pubkey for v in state.validators]
+//    committee_indices = [ValidatorIndex(all_pubkeys.index(pubkey)) for pubkey in state.current_sync_committee.pubkeys]
+//    for participant_index, participation_bit in zip(committee_indices, sync_aggregate.sync_committee_bits):
+//        if participation_bit:
+//            increase_balance(state, participant_index, participant_reward)
+//            increase_balance(state, get_beacon_proposer_index(state), proposer_reward)
+//        else:
+//            decrease_balance(state, participant_index, participant_reward)
+func ProcessSyncAggregate(state state.BeaconStateAltair, sync *ethpb.SyncAggregate) (state.BeaconStateAltair, error) {
+	currentSyncCommittee, err := state.CurrentSyncCommittee()
+	if err != nil {
+		return nil, err
+	}
+	if currentSyncCommittee == nil {
+		return nil, errors.New("nil current sync committee in state")
+	}
+	committeeKeys := currentSyncCommittee.Pubkeys
+	votedKeys := make([]bls.PublicKey, 0, len(committeeKeys))
+	votedIndices := make([]types.ValidatorIndex, 0, len(committeeKeys))
+	didntVoteIndices := make([]types.ValidatorIndex, 0, len(committeeKeys))
+	// Verify sync committee signature.
+	for i := uint64(0); i < sync.SyncCommitteeBits.Len(); i++ {
+		vIdx, exists := state.ValidatorIndexByPubkey(bytesutil.ToBytes48(committeeKeys[i]))
+		// Impossible scenario.
+		if !exists {
+			return nil, errors.New("validator public key does not exist in state")
+		}
+
+		if sync.SyncCommitteeBits.BitAt(i) {
+			pubKey, err := bls.PublicKeyFromBytes(committeeKeys[i])
+			if err != nil {
+				return nil, err
+			}
+			votedKeys = append(votedKeys, pubKey)
+			votedIndices = append(votedIndices, vIdx)
+		} else {
+			didntVoteIndices = append(didntVoteIndices, vIdx)
+		}
+	}
+	ps := helpers.PrevSlot(state.Slot())
+	d, err := helpers.Domain(state.Fork(), helpers.SlotToEpoch(ps), params.BeaconConfig().DomainSyncCommittee, state.GenesisValidatorRoot())
+	if err != nil {
+		return nil, err
+	}
+	pbr, err := helpers.BlockRootAtSlot(state, ps)
+	if err != nil {
+		return nil, err
+	}
+	sszBytes := p2pType.SSZBytes(pbr)
+	r, err := helpers.ComputeSigningRoot(&sszBytes, d)
+	if err != nil {
+		return nil, err
+	}
+	sig, err := bls.SignatureFromBytes(sync.SyncCommitteeSignature)
+	if err != nil {
+		return nil, err
+	}
+	if !sig.Eth2FastAggregateVerify(votedKeys, r) {
+		return nil, errors.New("could not verify sync committee signature")
+	}
```

Comment:
This looks like it could be abstracted into a helper method.

We have similar methods in the helper pkg: ComputeDomainAndSign, ComputeDomainVerifySigningRoot, VerifySigningRoot, etc